lib/Core/Executor.cpp

```diff
@@ -3942,60 +3942,129 @@ void Executor::executePersistentMemoryFlush(ExecutionState &state,
                                             ref<Expr> address) {
   address = optimizer.optimizeExpr(address, true);
 
-  ObjectPair op;
-  bool success;
+  ref<Expr> rangeBegin = memory->alignToCache(address);
+  ref<Expr> rangeEnd = AddExpr::create(rangeBegin,
+                                       memory->getCacheAlignmentExpr());
+
+  ResolutionList rl;
   solver->setTimeout(coreSolverTimeout);
-  if (!state.addressSpace.resolveOne(state, solver, address, op, success)) {
-    address = toConstant(state, address, "resolveOne failure");
-    success = state.addressSpace.resolveOne(cast<ConstantExpr>(address), op);
-  }
+  bool incomplete = state.addressSpace.resolveRange(state, solver,
+                                                    rangeBegin, rangeEnd,
+                                                    rl, 0, coreSolverTimeout);
   solver->setTimeout(time::Span());
 
-  if (success) {
-    const MemoryObject *mo = op.first;
-    const ObjectState *os = op.second;
+  // XXX bgreeves debugging
+  std::string msg;
+  llvm::raw_string_ostream out(msg);
+  out << "ResolutionList[size=" << rl.size() << "]: ";
+  for (ResolutionList::iterator i = rl.begin(), ie = rl.end(); i != ie; ++i) {
+    const MemoryObject *mo = i->first;
+    out << mo->name << "[";
+    out.write_hex(mo->address);
+    out << "] ";
+  }
+  klee_warning("%s", out.str().c_str());
+
+  if (incomplete) {
+    terminateStateEarly(state, "Query timed out (resolve).");
+  } else if (rl.empty()) {
+    terminateStateOnError(state, "memory error: out of bound pointer", Ptr,
+                          NULL, getAddressInfo(state, address));
+    return;
+  }
+
+  auto doFlush = [&](ExecutionState &state, const MemoryObject *mo,
```

Comment:
This function is long enough that it would be more readable to separate this out into a different `Executor` method with a docstring and all that jazz.